Use SVG-based toolbar bitmaps in webview sample

PBfordev opened the pull request:
## Introduction
This is a follow-up to [ticket 19287](https://trac.wxwidgets.org/ticket/19287). I have attempted to create SVG files for the four bitmaps used in the webview sample (back, forward, stop, refresh), as simple black outlines. The SVG for the menu is [wxWidgets logo](http://wxwidgets.org/downloads/logos/blocks.svg). This is how the bitmaps look now (on display with 125% scaling):
![webview-sample-toolbar-at125pct](https://user-images.githubusercontent.com/12495521/148832742-9eaab5ed-cfb4-4b9c-8972-c2cf892a97ad.png)

I can use only Windows, so the code was tested only there.

## Questions
### Q1. How to load the SVGs?
The SVGs are currently loaded from files. I do not like this but I did not know how to do it better. The 4 new icons are ~2 kB in total (I cut their SVGs to the bone) but wxlogo is ~12 kB, so I did not want to embed the SVGs in the source file. 
Additionally, when adding the SVGs to the sample bakefile and checking how it is done looking at the samples, I have noticed that not all samples have their data files in their bakefiles? I also do not know if I added them to bakefile correctly.
There are SVGs in the toolbar sample folder (`vector` folder) but they do not seem to be used in the sample (only the embedded one)?

### Q2. What to do when NanoSVG is not available?
Vadim wrote "I don't think we need `wxHAS_SVG` test, or at least not the fallback code, i.e. it would be fine to not show any icons if SVG is not supported." However, `wxToolbar::AddTool()` requires a valid `wxBitmap(Bundle)`. There is `wxTB_NOICONS` but the docs reads "Note that the wxMSW native toolbar ignores wxTB_NOICONS style."  Not sure what that means but when the style flag is used there are [no bitmaps shown](https://i.imgur.com/7A2eqsp.png) on Win10 and there is no assert when passing an empty `wxBitmapBundle`...

## Known issues
The issues may be somehow caused by the SVG files I created as I know nothing about SVG.

### I1. Drawings are ugly
The bitmaps are ugly, but I do not think I can really do better, hopefully someone else can.

### I2. MSW toolbar refuses to disable the bitmaps
For some reason, the bitmaps created from SVGs are not grayed-out when the command is disabled (but when I used green or red color for drawing, the disabled ones were drawn black). My uneducated guess is this may be caused by the bitmap type (DIB vs DDB), color depth/bitmaps having alpha (most likely culprit?), or something entirely else. For now, I did not investigate further.

### I3. The toolbar bitmap sizing is odd
The screenshot in the Introduction was taken from the primary display on a computer with two displays, the primary having DPI at 125% and the secondary having default 100%. The SVGs all have the same page size (based on wxlogo.svg) and the drawing is centered on the page to add padding around it.
When moving the window from 125% to 100% DPI display, the bitmaps seem to become larger (taller than the URL text control now):
![webview-sample-toolbar-at100from125pct](https://user-images.githubusercontent.com/12495521/148835513-151a988f-6d0c-4093-91cf-59406ca89559.png)
When I set DPI on the primary display to 100% as well, it looks like this on the primary display (probably the best look?):
![webview-sample-toolbar-at100pct](https://user-images.githubusercontent.com/12495521/148835600-9d4b2a9f-7135-4751-a4de-8c1da90b2955.png)

So in each of the three screenshots here, the bitmap size compared to the URL control is different?

### I4. The toolbar bitmaps are not centered
This is noticeable even when the text is used but perhaps it can be better seen with the labels off:
![webview-sample-toolbar-at125pct-detail](https://user-images.githubusercontent.com/12495521/148841797-c69e8d34-81ab-4aa2-b0ba-d758a3b3950d.png)
However, this may be caused by the padding, since this seemed OK when there was no padding at all, see [here](https://trac.wxwidgets.org/ticket/19287#comment:3).

PBfordev committed:
Use SVG-based toolbar bitmaps in webview sample

In webview sample toolbar use wxBitmapBundles created
from SVG files instead of XPM bitmaps.

PBfordev commented:
I will push the whitespace fix for the issue that failed the check only when there are other commits, to not needlessly trigger CI.

PBfordev committed:
Remove extra space in CMakefile for samples

PBfordev committed:
Restore accidentally removed include for the sample XPM icon

PBfordev commented on samples/webview/webview.cpp:
```diff
@@ -309,31 +304,32 @@ WebFrame::WebFrame(const wxString& url) :
+    if ( !sizeBitmap.IsFullySpecified() )
+        sizeBitmap.Set(32, 32);
+
+    wxBitmapBundle back    = wxBitmapBundle::FromSVGFile("webview_back.svg", sizeBitmap);
```
`wxBitmapBundle::FromSVGFile()` calls should have been guarded with `ifndef wxHAS_SVG` and the variables declared before the guarded block.

PBfordev commented on samples/webview/webview.cpp:
```diff
@@ -309,31 +304,32 @@ WebFrame::WebFrame(const wxString& url) :
+    wxSize sizeBitmap = wxArtProvider::GetNativeSizeHint(wxART_TOOLBAR);
+
+    if ( !sizeBitmap.IsFullySpecified() )
+        sizeBitmap.Set(32, 32);
```
Just curious, why was 32 chosen in the original code (regardless of the platform and DPI)?

vadz replied:
I don't think there is any deep reason for this, it's just a reasonable default. For most platforms `GetNativeSizeHint()` should return valid values, I think.

vadz commented:
I don't have any perfect answer for how to load SVGs. We can include them as resources under MSW and/or use bin2c and include them in the source code itself or load from external files. For the samples I think the latter is the best because it's simple and the least magical. They do need to be listed as "data" in the bkl files, if there are any samples for which this is not done, it could be a mistake -- which ones are those?

Unfortunately I can help even less with the other issues, i.e. I have no idea why are the bitmaps not disabled and not centered. I'll try to debug it later.

The only thing I can say is that they're not ugly at all IMO and are definitely an improvement compared to the currently used bitmaps. Thanks!

PBfordev committed:
Update the SVGs

The drawings are the same but the dimensions, viewbox,
and formatting are different.

PBfordev committed:
Do not use SVGs when wxHAS_SVG is not defined

vadz commented:
Thanks again for your work on this!

I can answer at least one of your questions: using `wxArtProvider::GetNativeSizeHint()` as the default bundle size is indeed wrong, thanks for noticing it! We need to use `ToDIP()` with this value, which is in logical pixels, to convert it to DIPs under MSW (elsewhere LP == DIP already). It's annoying to have to do it, but it seems correct to both use DIPs for the default bundle size (which shouldn't depend on DPI) and LPs for the return value of `wxArtProvider` function, which means that we really need to use this function here.

Concerning the disabled bitmaps, do I understand correctly that the problem is just that MSW doesn't change black pixels when disabling, i.e. the disabled appearance of the monochrome bitmaps is simple the same as normal one? If so, could we just use colour for them, e.g. make them blue or something?

Another tiny question concerning them is that I think it might be better to put wx logo into `samples/sample.svg`, as it could probably be reused in the other samples too, what do you think?

Other than that I'd like to merge this, but probably with a squash merge as it doesn't seem worth keeping the individuals commits her -- or would you like to preserve them? Please let me know if so.

PBfordev committed:
Update SVGs

The drawings are now all colored and have 32x32 size

Additionally, the viewbox is reset (originally used for padding).

PBfordev commented:
>I can answer at least one of your questions: using wxArtProvider::GetNativeSizeHint() as the default bundle size is indeed wrong, thanks for noticing it! We need to use ToDIP()...

I may be missing something (again), but I think this may not be always correct. For example, in a situation where the window calling `wxArtProvider::GetNativeSizeHint()` is on a screen with 125% DPI but the window actually used by `GetNativeSizeHint()` (i.e, `wxApp::GetMainTopWindow()`) is on a screen with 100% DPI, we get `19` instead of `30` pixels? As `wxArtProvider::GetNativeSizeHint()` is not DPI-agnostic, perhaps it should also take a window as an argument, which would be used, when available, instead of `wxApp::GetMainTopWindow()`. In other words, similar to `wxSystemSettings::GetMetric()`?

> Concerning the disabled bitmaps, do I understand correctly that the problem is just that MSW doesn't change black pixels when disabling, i.e. the disabled appearance of the monochrome bitmaps is simple the same as normal one? If so, could we just use colour for them, e.g. make them blue or something?

Yes, the black is left untouched in MSW toolbar disabled bitmaps.

I colored the SVGs, changed them to be all 32x32, and also reset the viewbox originally used for padding. This makes the simple-looking bitmaps quite ugly at large size, but I think padding should not be done at the SVG level. The UI control should take care of that, unfortunately, even if at least MSW toolbar supports padding, wxToolBar lacks that. So it looks like this now:
![webview-sample-toolbar-color](https://user-images.githubusercontent.com/12495521/149555623-907342a9-b538-4ba0-a2ff-65ac5438c36c.png)


BTW, when playing with the toolbar sample, I noticed a bug, not sure if it is in the sample or `wxToolBar`. When in the sample one switches toolbar with menu Toolbar from "View both" to "View icons", the toolbar buttons become small (matching the 16x15 bitmap size mentioned both in `wxToolBar`'s and Win32 toolbar docs and/or matching the combobox height?) and only the top left part of the button bitmap is shown. This fixes itself after another switching view to Text and back but can be also observed in Both mode. I have also noticed that the background of the comobobox pop-up window is initially black after switching View mode?
![toolbar-sample-bug](https://user-images.githubusercontent.com/12495521/149554040-4ba8f7eb-fcae-4ef0-9310-4ce83a1f6dd6.png)

> Another tiny question concerning them is that I think it might be better to put wx logo into samples/sample.svg, as it could probably be reused in the other samples too, what do you think?

In general, this is a good idea and I considered it before. However, it will make the sample code more complicated, to make sure the sample can always find the file. There is no problem when the sample is ran from the folder where its executable is, but will not work when the sample is ran from an IDE such as MSVS where the CWD is the folder with the project and not the executable.
We would probably need to do something like
```
wxPathList pathlist;

pathlist.Add("."); 
pathlist.Add("../samples");
logoPath = wxFileName(pathlist.FindValidPath("sample.svg")).GetAbsolutePath();
```

> Other than that I'd like to merge this, but probably with a squash merge as it doesn't seem worth keeping the individuals commits her -- or would you like to preserve them? Please let me know if so.

Yes, please squash the commits. But if possible, I would like if someone checked how it looks on GTK/MacOS before merging...

vadz commented:
Thanks for the update! I've tested under Linux (in high DPI) and things looks perfectly fine, so I'm going to merge this.

We do need to do something about the `wxArtProvider::GetSizeHint()` problem and I've created #22022 for it.

I didn't have time to look at the bug with switching from icons to text+icons. If the sample calls `Realize()` after this change (and I think it does), it's almost certainly a bug in wxMSW too and ideally should be fixed. I don't know if it's high DPI-specific too, i.e. does this happen at 100% DPI as well?

Finally, I'll keep the logo where it is for now and maybe we'll move it, and add something similar to your snippet using `wxPathList` to this sample, later if/when we need it somewhere else.

Thanks again!

TcT2k commented:
Here is a screen shot from the sample running on macOS:
<img width="1017" alt="Screenshot 2022-01-24 at 15 07 13" src="https://user-images.githubusercontent.com/5075894/150798082-1fc551c0-c1e9-4a91-abfd-a0ea9e9cf138.png">

I've also tested on windows:
![image](https://user-images.githubusercontent.com/5075894/150798454-d500f75a-2458-4de5-b7c7-f3b53e48d883.png)

While I think it's a definitive improvement over the pixelated bitmaps, I think the SVGs could use with a little padding around the edges.

vadz commented:
Thanks for testing and, indeed, they look huge under MSW. Which DPI scaling is used here?

TcT2k commented:
> Thanks for testing and, indeed, they look huge under MSW. Which DPI scaling is used here?

That is 350% on Windows 11. Haven't tested on another Windows machine with lower scale yet.

vadz commented:
Ah, so it's normal that they look so huge. FWIW they still look biggish, but IMHO acceptably so in 100% and 200% DPI that I've tested with.

I wouldn't mind adding padding around them anyhow, of course, the trouble is that I'm not even sure how to do this. I should spend some time with Inkskape, I guess...

PBfordev commented:
The lack of padding is an issue, but as I wrote above, I did not think it should be done at the SVG level (see the lack of `wxToolbar::SetPadding()`). However, it seems I was wrong and I will fix the images. I think I should create them at size 20 centered inside a page of 24.

Additionally, the recently discussed problems with `wxArtProvider::GetNativeSizeHint()` may be another factor contributing to overly large toolbar button size.

The current problems aside, I do not understand why it returns `24` for toolbar on MSW as the baseline. Both wxWidgets and Microsoft documentation state that the default toolbar bitmap size is `16`x`15`/`16`.
